parsers/SV.py

```diff
@@ -0,0 +1,198 @@
+
+url = 'http://estadistico.ut.com.sv/OperacionDiaria.aspx'
+
+generation_map = {
+                  0: 'biomass',
+                  1: 'geothermal',
+                  2: 'hydro',
+                  3: 'interconnection',
+                  4: 'thermal',
+                  5: 'solar',
+                  'datetime': 'datetime'
+                  }
+
+def get_data(session = None):
+    """
+    Makes a get request to data url.  Parses the response then makes a post request to the same url using
+    parsed parameters from the get request.
+    Returns a requests response object.
+    """
+
+    s = session or requests.Session()
+    pagereq = s.get(url)
+
+    soup = BeautifulSoup(pagereq.content, 'html.parser')
+
+    #Find and define parameters needed to send a POST request for the actual data.
+    viewstategenerator = soup.find("input", attrs = {'id': '__VIEWSTATEGENERATOR'})['value']
+    viewstate = soup.find("input", attrs = {'id': '__VIEWSTATE'})['value']
+    eventvalidation = soup.find("input", attrs = {'id': '__EVENTVALIDATION'})['value']
+    DXCss = '1_33,1_4,1_9,1_5,15_2,15_4'
+    DXScript = '1_232,1_134,1_225,1_169,1_187,15_1,1_183,1_182,1_140,1_147,1_148,1_142,1_141,1_143,1_144,1_145,1_146,15_0,15_6,15_7'
+    callback_param_init = 'c0:{"Task":"Initialize","DashboardId":"OperacionDiaria","Settings":{"calculateHiddenTotals":false},"RequestMarker":0,"ClientState":{}}'
+
+
+    postdata = {'__VIEWSTATE': viewstate,
+                '__VIEWSTATEGENERATOR': viewstategenerator,
+                '__EVENTVALIDATION': eventvalidation,
+                '__CALLBACKPARAM': callback_param_init,
+                '__CALLBACKID':'ASPxDashboardViewer1',
+                'DXScript': DXScript,
+                'DXCss': DXCss
+               }
+
+    datareq = s.post(url, data=postdata)
+
+    return datareq
+
+
+def data_parser(datareq):
+    """
+    Accepts a requests response.text object.  Slices the object down to a smaller size then converts
+    to usable json.  Loads the data as json then finds the 'result' key.  Uses regex to find the start
+    and endpoints of the actual data.  Splits the data into datapoints then cleans them up for processing.
+    Returns a list of lists.
+    """
+
+    double_json = datareq.text[len('0|/*DX*/('):-1]
+    double_json = double_json.replace('\'', '"')
+    data = (json.loads(double_json))
+    jsresult = data['result']
+
+    startpoints = [m.end(0) for m in re.finditer('"DataItem0":0},"Data":{"', jsresult)]
+    endpoints = [m.start(0) for m in re.finditer('"KeyIds"', jsresult)]
+
+    sliced = jsresult[startpoints[1]:endpoints[2]-4]
+    chopped = sliced.split( ',"')
+
+    diced = []
+    for item in chopped:
+        item = item.replace("}", "")
+        np = item.split('":')
+        diced.append(np[0::2])
+
+    #literal_eval is used to convert a string representation of a list into an actual list.
+    clean_data = []
+    for item in diced:
+        j = ast.literal_eval(item[0])
```

Comment:
Note: you can also use `json.loads` here instead of `ast.literal_eval`. It works the same in this case but might be a bit safer in case of rogue data.

Reply:
Yep I didn't think of that but json instead of eval makes more sense.